apisix/plugins/google-cloud-logging.lua

```diff
@@ -0,0 +1,326 @@
+        },
+    })
+
+    if err then
+        return nil, "failed to write log to google, " .. err
+    end
+
+    if res.status ~= 200 then
+        return nil, res.body
+    end
+
+    return res.body
+end
+
+
+local function get_auth_config(config)
+    if auth_config_cache then
+        return auth_config_cache
+    end
+
+    if config.auth_config then
+        auth_config_cache = config.auth_config
+        return auth_config_cache
+    end
+
+    if not config.auth_file then
+        return nil, "configuration is not defined"
+    end
+
+    local file_content, err = core.io.get_file(config.auth_file)
+    if not file_content then
+        return nil, "failed to read configuration, file: " .. config.auth_file .. " err:" .. err
+    end
+
+    local config_data
+    config_data, err = core.json.decode(file_content)
+    if not config_data then
+        return nil, "config parse failure, data: " .. file_content .. " , err: " .. err
+    end
+
+    auth_config_cache = config.auth_config
+    return auth_config_cache
+end
+
+
+local function get_logger_buffer(conf, ctx)
+    local oauth_client = google_oauth:new(auth_config_cache)
+
+    local process = function(entries)
+        return send_to_google(oauth_client, entries)
+    end
+
+    local config = {
+        name = conf.name or plugin_name,
+        retry_delay = conf.retry_delay,
+        batch_max_size = conf.batch_max_size,
+        max_retry_count = conf.max_retry_count,
+        buffer_duration = conf.buffer_duration,
+        inactive_timeout = conf.inactive_timeout,
+        route_id = ctx.var.route_id,
+        server_addr = ctx.var.server_addr,
+    }
+
+    local buffer, err = batch_processor:new(process, config)
+
+    if not buffer then
+        return nil, "error when creating the batch processor: " .. err
+    end
+
+    return buffer
+end
+
+
+local function get_utc_timestamp()
+    ngx_update_time()
+    local now = ngx_now()
+    local second = math_floor(now)
+    local millisecond = math_floor((now - second) * 1000)
+    return os_date("!%Y-%m-%dT%T.", second) .. core.string.format("%03dZ", millisecond)
+end
+
+
+local function get_logger_entry(conf, ctx)
+    local auth_config, err = get_auth_config(conf)
+    if err or not auth_config.project_id or not auth_config.private_key then
+        return nil, "failed to get google authentication configuration, " .. err
+    end
+
+    local entry = log_util.get_full_log(ngx, conf)
+    local google_entry = {
+        httpRequest = {
+            requestMethod = entry.request.method,
+            requestUrl = entry.request.url,
+            requestSize = entry.request.size,
+            status = entry.response.status,
+            responseSize = entry.response.size,
+            userAgent = entry.request.headers and entry.request.headers["user-agent"],
+            remoteIp = entry.client_ip,
+            serverIp = entry.upstream,
+            latency = tostring(core.string.format("%0.3f", entry.latency / 1000)) .. "s"
```

Comment:
Looks like this is more accurate:  https://github.com/apache/apisix/blob/5394dce6fb9dde973baee41a31107d99ab2ca1f3/apisix/utils/log-util.lua#L180?

Reply:
The format is google requirements: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry